frontend/src/metabase/query_builder/components/notebook/lib/steps.ts

```diff
@@ -1,173 +1,222 @@
-    revert: (legacyQuery, itemIndex, query, stageIndex) => {
-      const reverted = Lib.clearOrderBys(query, stageIndex);
-      return convertStageQueryToLegacyStageQuery(
-        reverted,
-        legacyQuery,
-        stageIndex,
+    valid: (query, stageIndex) => {
+      const hasAggregations = Lib.aggregations(query, stageIndex).length > 0;
+      const hasBreakouts = Lib.breakouts(query, stageIndex).length > 0;
+
+      if (hasAggregations && !hasBreakouts) {
+        return false;
+      }
+
+      return (
+        hasData(query) && (stageIndex === 0 || hasAnyClauses(query, stageIndex))
       );
     },
-
-    // Order-bys can only be added from the notebook editor with MLv2.
-    // MLv2 guarantees that a query is valid at any given point of time,
-    // so we can skip order-bys cleaning.
-    clean: query => query,
+    active: (query, stageIndex) => {
+      return Lib.orderBys(query, stageIndex).length > 0;
+    },
+    revert: (query, stageIndex) => {
+      return Lib.clearOrderBys(query, stageIndex);
+    },
   },
   {
     type: "limit",
-    valid: query =>
-      query.hasData() &&
-      (!query.hasAggregations() || query.hasBreakouts()) &&
-      (!query.sourceQuery() || query.hasAnyClauses()),
-    active: (legacyQuery, itemIndex, query, stageIndex) =>
-      Lib.hasLimit(query, stageIndex),
-    revert: (legacyQuery, itemIndex, query, stageIndex) => {
-      const reverted = Lib.limit(query, stageIndex, null);
-      return convertStageQueryToLegacyStageQuery(
-        reverted,
-        legacyQuery,
-        stageIndex,
+    valid: (query, stageIndex) => {
+      const hasAggregations = Lib.aggregations(query, stageIndex).length > 0;
+      const hasBreakouts = Lib.breakouts(query, stageIndex).length > 0;
+
+      if (hasAggregations && !hasBreakouts) {
+        return false;
+      }
+
+      return (
+        hasData(query) && (stageIndex === 0 || hasAnyClauses(query, stageIndex))
       );
     },
-    clean: query => query,
+    active: (query, stageIndex) => {
+      return Lib.hasLimit(query, stageIndex);
+    },
+    revert: (query, stageIndex) => {
+      return Lib.limit(query, stageIndex, null);
+    },
   },
 ];
 
+const hasData = (query: Lib.Query): boolean => {
+  const databaseId = Lib.databaseID(query) !== null;
+  return databaseId !== null;
+};
+
+const hasAnyClauses = (query: Lib.Query, stageIndex: number): boolean => {
+  const hasAggregations = Lib.aggregations(query, stageIndex).length > 0;
+  const hasBreakouts = Lib.breakouts(query, stageIndex).length > 0;
+  const hasJoins = Lib.joins(query, stageIndex).length > 0;
+  const hasExpressions = Lib.expressions(query, stageIndex).length > 0;
+  const hasFilters = Lib.filters(query, stageIndex).length > 0;
+  const hasOrderBys = Lib.orderBys(query, stageIndex).length > 0;
+  const hasLimits = Lib.hasLimit(query, stageIndex);
+  const hasFields = Lib.fields(query, stageIndex).length > 0;
+
+  return (
+    hasJoins ||
```

Comment:
it would be nice to have the same order as in `const` block

Reply:
9f8a79ad01777317e83a5c8f95817c3f88635488